### PR TITLE
Add valuable ground item notifications

### PR DESCRIPTION
**Pull request by unmrshl**
Closes https://github.com/runelite/runelite/issues/3643

notifications

Comments/criticisms are welcome

### CODE REVIEW

**Comment by deathbeam** [comment]
I dont think there should be dropdown for different "thresholds". It should throw notification if the item is either highlighted or "colored".

**Comment by c5gaspar** [comment]
I think there's an argument for notifications being "thresholded" (see below), and I do agree that's a great idea to include highlighted items too (if not already). 

When I'm actively engaged in the game (client in-focus), I'll pick up any items > 1K (this is what my "low value" is set to). Whereas if I'm working on something in the background (client out-of-focus), I don't want to be bothered every time that a 1K item drops, I'd rather have it set to something higher (e.x., 10K). In the event that the notification isn't "thresholded", I would have to switch my 'low value price' everytime that I didn't want to receive unnecessary notifications. 

This could be proactively avoided by the thresholds @marshdevs has implemented.

**Comment by marshdevs** [comment]
I hadn't even thought about highlighted items, that's a clear miss and I'll update to include those. 

I do however think a good argument has been made above for for selectively notifying at different colors.

**Comment by Unravl** [comment]
this is great, i was just about to work on something like this!

**Comment by deathbeam** [comment]
Needs rebase. also, probably instead of toggle, add OFF in the dropdown

**Comment by marshdevs** [comment]
Rebased with new version of ground items, removed the toggle option, and set the default threshold to 'OFF'.

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsOverlay.java [hunk 1]
```diff
@@ -215,6 +216,15 @@ public Dimension render(Graphics2D graphics)
 				}
 			}
 
+			final int dropValue = item.getGePrice() > 0 ? item.getGePrice() : item.getHaPrice();
+			final ValueThreshold costThreshold = color.equals(config.highlightedColor()) ? ValueThreshold.HIGHLIGHTED : getCostThreshold(dropValue);
+
+			if (costThreshold != null && costThreshold.isGreaterThan(config
+					.notifierThreshold()))
+			{
+				plugin.notifyValuableItem(item, costThreshold);
```
[the remark this answers is not in the document]
Comment (marshdevs) [reply runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsOverlay.java]: Addressed these in newest commit. I also made a change which involves passing a config object around. If this is poor practice, I can revert

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/config/ValueThreshold.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	{
+		return name;
+	}
+
+	public boolean isGreaterThan(ValueThreshold threshold)
+	{
+		return value >= threshold.value;
+	}
+
+	public static ValueThreshold getThreshold(int cost, GroundItemsConfig config)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/config/ValueThreshold.java]: I would probably make this work in similar way as colors do atm, so just populate some list and iterate it.

**Comment by marshdevs** [comment]
Made requested changes

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/config/ValueThreshold.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package net.runelite.client.plugins.grounditems.config;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/config/ValueThreshold.java]: Add license header

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 4]
```diff
@@ -342,29 +356,37 @@ private void reset()
 		// Cache colors
 		priceChecks.clear();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Remove this empty line

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 5]
```diff
@@ -279,6 +285,14 @@ void checkItems()
 					{
 						groundItem.setHeight(itemLayer.getHeight());
 						groundItems.add(groundItem);
+
+						ValueThreshold valueThreshold = getThreshold(groundItem.getName(), groundItem.getGePrice(), groundItem.getHaPrice());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Mark this final

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 6]
```diff
@@ -525,4 +566,36 @@ public void onFocusChanged(FocusChanged focusChanged)
 			setHotKeyPressed(false);
 		}
 	}
+
+	public void notifyValuableItem(GroundItem item, ValueThreshold threshold)
+	{
+		final Player local = client.getLocalPlayer();
+		StringBuilder notificationString = new StringBuilder();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: final

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 7]
```diff
@@ -525,4 +566,36 @@ public void onFocusChanged(FocusChanged focusChanged)
...
+
+		notificationString.append(" drop: ")
+			.append(item.getName())
+			.append("");
+
+		if (item.getQuantity() > 1)
+		{
+			notificationString.append(" x ").append(item.getQuantity());
+		}
+		notifier.notify("[" + local.getName() + "] received " + notificationString.toString() + "!");
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: empty line above

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: I would also use here same format as overlay uses for displaying an item (and use some shared logic)

**Comment by marshdevs** [comment]
Oof, been a while. Made requested changes a while ago, and just rebased today.

**Comment by jold1908** [comment]
Is this in runelite atm? i cant find it

**Comment by unmrshl** [comment]
This was never approved. If there’s demand for it, I’ll rebase.

**Comment by deathbeam** [comment]
I still dont really like how this is done. Current highlighting code supports variable amount of price thresholds, while your changes for sending notifications for them does not (e.g you need to add new field to enum etc). I dont think this can be made flexible tho, so maybe just notifying on explicitely highlighted items is enough?

**Comment by jold1908** [comment]
I would really want this feature. Im afking alot when on school and dont have runelite running in the back. I miss drops often.

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/config/ValueThreshold.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+	@Override
+	public String toString()
+	{
+		return name;
+	}
+
+	public boolean isGreaterThan(ValueThreshold threshold)
+	{
+		return value >= threshold.value;
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/config/ValueThreshold.java]: You can just test `ordinal` here

**Comment by Adam-** [comment]
This approach seems fine to me. Maybe you could check the config doesn't have it off before running the check to see what threshold it is in, though.

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 9]
```diff
@@ -206,6 +217,16 @@ public void onItemSpawned(ItemSpawned itemSpawned)
 		{
 			existing.setQuantity(existing.getQuantity() + groundItem.getQuantity());
 		}
+
+		if (!config.notifierThreshold().equals(ValueThreshold.OFF))
```
Comment (NathenSample) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Reversing the condition and simply returning early would make this easier to understand and save on nesting `if`'s

**Comment by deathbeam** [comment]
I still disagree, the way how the thresholds are done atm is that they are easily and dynamically expandable in future, this PR is basically throwing that away

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 10]
```diff
@@ -484,4 +530,51 @@ public void onFocusChanged(FocusChanged focusChanged)
...
+		}
+
+		notificationStringBuilder.append(" drop: ")
+			.append(item.getName());
+
+		if (item.getQuantity() > 1)
+		{
+			notificationStringBuilder.append(" x ").append(item.getQuantity());
+
+			if (item.getQuantity() >= MAX_QUANTITY)
```
Comment (NathenSample) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Abstracting this logic to its own method would also allow for the removal of the `if else` construct by utilizing returns

**Comment by unmrshl** [comment]
@deathbeam Can you explain what you mean by a variable amount of price thresholds? This doesn't look variable to me.

https://github.com/runelite/runelite/blob/21d00e4ea8c666f81a2e1819a7ca9bc33d61fc3d/runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java#L306-L329

**Comment by deathbeam** [comment]
The priceCheck array can contain any amount of thresholds

**Comment by brace110** [comment]
The demand for this feature is very real, I hope this request hasn't been forgotten ?

**Comment by deathbeam** [comment]
I still think just notifying on highlighted items is the way to go.

**Comment by deathbeam** [comment]
And what I meant by it being not flexible enough, is that you cant dynamically add to enum, but you can dynamically add to that map. E.g if we wanted to support variable amount of thresholds with some custom config type, the entire enum would have to be dropped.

**Comment by unmrshl** [comment]
Gotcha. By highlighted, do you mean explicitly highlighted (the default purple color)? Or all items colored anything other than white/hidden?

**Comment by brace110** [comment]
@deathbeam I actually really like the select dropdown that allows you to pick the threshold value. I would see a need for certain people to have the notification at different rates. But I see your point about customizability and future proofing the code base. Is there any way we at least could get the first option and later refactor it to include your points? Since now we have neither..

**Comment by unmrshl** [comment]
Removed ValueThreshold. Updated notification to an on/off toggle. Only allows notifications for items in the highlighted list.

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 11]
```diff
@@ -90,6 +93,9 @@
 		.trimResults();
 
 	private static final Joiner COMMA_JOINER = Joiner.on(",").skipNulls();
+	// The game won't send anything higher than this value to the plugin -
+	// so we replace any item quantity higher with "Lots" instead.
+	private static final int MAX_QUANTITY = 65535;
```
[the remark this answers is not in the document]
Comment (unmrshl) [reply runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: (int) Character.MAX_VALUE = 65535

[the remark this answers is not in the document]
Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Regardless of how it's expressed, this is probably better defined in an API file for widespread use.

**Pull request merged**